### PR TITLE
Docker: Migrate from MySQL to MariaDB to allow for better Apple M1 compatibility

### PR DESCRIPTION
**Pull request by bisko**
This PR is a redo of #19237

This change migrates the database engine from MySQL to MariaDB, because MySQL images do not support the arm/v8 architecture the new Apple Sillicon computers use.

Seems that the MySQL images built on Docker Hub are only for the `amd64` architecture and can't run on the ARM architecture. 

MariaDB on the other hand has builds for more architectures and it should be a drop-in replacement for MySQL.

#### Changes proposed in this Pull Request:

* Upgrade the Docker database image from MySQL to MariaDB

#### Jetpack product discussion

N/a

#### Does this pull request change what data or activity we track or use?

N/a

#### Testing instructions:

* Stop Docker stack gracefully - either CTRL+C the running `jetpack docker up` (if it's not started in the detached state with `-d` argument) or run `jetpack docker down`
  * This is required, because the migration to MariaDB can't happen if the database is crashed 
* Backup your database folder from `tools/docker/data/jetpack_dev_mysql` 
* Apply branch
* `jetpack docker up`
* Monitor the logs for errors and see if the DB image didn't manage to start (it should start)
* Monitor the logs and see if the migration was automatically performed correctly and no errors were shown.
* Make sure your test site is working

### CODE REVIEW

**Review by jeherve — COMMENTED** [review]
It migrated well for me, seemed to work fine. It'd be nice if folks didn't have to run the migration on their own though, if we'd run the migration step for them when powering up the first time. Can you think of a way to do that?

You'll also want to update the docs in
https://github.com/Automattic/jetpack/blob/4cefb8405dfe643d9003a13cd4b3e26572800338/tools/docker/README.md

**Comment by kraftbj** [comment]
I agree that we should run the migration ourselves if possible on first run. Is there a way to programmatically determine if the upgrade is needed and run if so?

**Comment by samiff** [comment]
If we're not sure about automating the migration, would relying on documentation guidance suffice? Punted to 10.1 milestone.

**Comment by bisko** [comment]
@Automattic/ground-control I think this is in a state that's much closer to merging than before.

The new changes make a temporary container that starts, waits for the `db` container to come up and then tries to perform an upgrade. Then it exits out and shuts down the container.

A few things of note for this as I'm not sure how it will play with systems down the line:

* It adds yet another container to start up on every `jetpack docker up`.
  * Would this cause issues for any automated/CI flows that we have - resources/performance wise? I don't see issues on my side as startup is nearly instantaneous.
* It has a bit of hardcoded credentials in the script. Not sure if this is OK.
  * Refers to the DB container via an alias, as the container might be `jetpack_dev_db_1` or something else, depending on the environment. Not sure how to do a proper reference to the `db` container in the current compose context apart from using `links`.
* I am totally unsure if the inline documentation is sufficient in this case. Would love some pointers how to make it better.
* Are there enough fail state catches in the bash script. Any input here would be very wonderful and appreciated!

**Comment by jeherve** [comment]
> It adds yet another container to start up on every `jetpack docker up`
> Would this cause issues for any automated/CI flows that we have - resources/performance wise? I don't see issues on my side as startup is nearly instantaneous.

@brbrr Do you think you could take a look at this, so we're sure it doesn't break CI?

**Comment by brbrr** [comment]
I checked it earlier today. I also was a bit concerned about possible negative effect on E2E runs in CI, but After comparing it with other PRs, it seems that it does not affect performance.

**Review by jeherve — DISMISSED** [review]
This tests well for me. Let's give it a try with everyone :)

**Comment by anomiex** [comment]
So far the upgrade isn't working right for me on Linux. It winds up dying with complaints about being unable to access /var/log/mysql/slow.log. Continuing to try to debug.

**Review by anomiex — CHANGES_REQUESTED** [review]
Ok, the problem is as follows:

* Docker creates the volumes for `./data/${COMPOSE_PROJECT_NAME}_mysql:/var/lib/mysql` and `./logs/${COMPOSE_PROJECT_NAME}/mysql:/var/log/mysql` owned by root.
* The mysql and mariadb images `chown` the `/var/lib/mysql` to the necessary user, but do not do so for `/var/log/mysql`.
* When running normally, mysql and mariadb both seem ok with not being able to actually write the logs. But the upgrade process chokes on it.
* I guess Macs don't have the same permissions structure with respect to docker, so they don't run into this for whatever reason.

As for fixing this permissions issue, the easiest thing to do might be to just have the new updater script do the `chown`. Or we could see about mounting the logs directory at `/var/lib/mysql/logs` or something like that, so it gets caught in the `chown` the image does.

**Comment by bisko** [comment]
@anomiex I added the log directory to be mapped into the container and that might fix the issue. Since I can't fully reproduce on macOS as it's using different way to map paths, compared to Linux, can you give it a go and see if the last change fixes the issue you're seeing?

File: tools/docker/bin/docker-db-autoupgrade.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+# Automatically upgrade the MySQL/MariaDB databases when starting the container
+# This is required on every major database version bump.
+# The script will wait until we can connect to the database server in the main container,
+# run the upgrade command and then exit.
+
+CREDENTIALS_PARAMS="-uroot -p${MYSQL_ROOT_PASSWORD} -hdb"
+
+# Wait until server is up
+while ! mysql $CREDENTIALS_PARAMS ${MYSQL_DATABASE} -e "SELECT 1 FROM wp_posts LIMIT 1" > /dev/null 2>&1
```
Comment (anomiex) [inline tools/docker/bin/docker-db-autoupgrade.sh:11]: This loops forever if WordPress hasn't been installed.

You don't really even need to do a query, all we're doing is testing the connection so the command can just be "quit".
```suggestion
while ! mysql $CREDENTIALS_PARAMS ${MYSQL_DATABASE} -e 'quit' > /dev/null 2>&1
```

**Review by anomiex — COMMENTED** [review]
Still doesn't work for me. I pushed #21142 with some changes that do make it work (and takes a somewhat different approach to avoid the extra container too).

**Review by anomiex — APPROVED** [review]

**Pull request merged**